### PR TITLE
misc: Add deviceid to configuration and enhance switchconfig

### PR DESCRIPTION
**Pull request by ccuetom**
Adds `deviceid` to `configuration` and updates `switchconfig` so that it can act both as a decorator and a context manager.

### CODE REVIEW

File: devito/types/parallel.py [hunk 1]
```diff
@@ -252,7 +252,7 @@ class DeviceID(DeviceSymbol):
 
     @property
     def default_value(self):
-        return -1
+        return configuration['deviceid']
```
Comment (FabioLuporini) [inline devito/types/parallel.py:255]: We can return a `default_value` based on global state

I think the better way to go here would be overriding `_arg_values` (from `Scalar`)

Within the override, we check, in order:

* whether kwargs has a user-override for `deviceid`
* whether `configuration['deviceid']` is != `default_value`
* else return `default_value`

**Review by FabioLuporini — CHANGES_REQUESTED** [review]
The changes in switchconfig are solid -- thanks!

My only request here is to edit DeviceID (see comment in review)

File: devito/parameters.py [hunk 2]
```diff
@@ -7,6 +7,7 @@
 from devito.logger import info, warning
 from devito.tools import Signer, filter_ordered
 
+
```
Comment (georgebisbas) [inline devito/parameters.py:10]: ?

File: devito/operator/operator.py [hunk 3]
```diff
@@ -1186,7 +1186,9 @@ def parse_kwargs(**kwargs):
 
     # `allocator`
     kwargs['allocator'] = default_allocator(
-        '%s.%s' % (kwargs['compiler'].__class__.__name__, kwargs['language'])
+        '%s.%s.%s' % (kwargs['compiler'].__class__.__name__,
```
Comment (mloubout) [inline devito/operator/operator.py:1189]: Can we just make a `name` property in our Compilers so we don't have to do this

File: devito/types/parallel.py [hunk 4]
```diff
@@ -254,6 +254,14 @@ class DeviceID(DeviceSymbol):
     def default_value(self):
         return -1
 
+    def _arg_values(self, **kwargs):
+        if self.name in kwargs:
+            return {self.name: kwargs.pop(self.name)}
+        elif configuration['deviceid'] != self.default_value:
+            return {self.name: configuration['deviceid']}
+        else:
+            return {self.name: self.default_value}
```
Comment (mloubout) [inline devito/types/parallel.py:263]: why not just `{self.name: kwargs.pop(self.name, self.default_value)}` and have `default_value` return `configuration['deviceid']`

File: devito/parameters.py [hunk 5]
```diff
@@ -232,31 +225,38 @@ def init_configuration(configuration=configuration, env_vars_mapper=env_vars_map
...
             self.params = {k.replace('_', '-'): v for k, v in params.items()}
         else:
             self.params = {}
+        self.previous = {}
+
+    def __enter__(self, condition=True, **params):
+        self.previous = {}
+        for k, v in self.params.items():
+            self.previous[k] = configuration[k]
+            configuration[k] = v
```
[the remark this answers is not in the document]
Comment (ccuetom) [reply devito/parameters.py:242]: this has a slightly different effect, although I could write it more succinctly:

```
self.previous = {k: configuration[k] for k in self.params.keys()}
configuration.update(**self.params)
```

File: devito/parameters.py [hunk 6]
```diff
@@ -232,31 +225,38 @@ def init_configuration(configuration=configuration, env_vars_mapper=env_vars_map
...
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        for k, v in self.params.items():
+            try:
+                configuration[k] = self.previous[k]
+            except ValueError:
+                # E.g., `platform` and `compiler` will end up here
+                try:
+                    configuration[k] = self.previous[k].name
+                except AttributeError:
+                    super(Parameters, configuration).__setitem__(k, self.previous[k])
```
[the remark this answers is not in the document]
Comment (ccuetom) [reply devito/parameters.py:253]: Not exactly, this bypasses the `@_check_key_value` decorator and the `_update_functions` callback. The value we're setting back into the `configuration` has already gone through those and in some cases (e.g. compiler and platform) it has been transformed. So the previous checks and transformations need to be skipped.

File: devito/types/parallel.py (hunk 4 above)
Comment (ccuetom) [reply devito/types/parallel.py:263]: isn't that what the original code did? I thought the idea was to keep `default_value` and `configuration['deviceid']` separate?

**Review by FabioLuporini — APPROVED** [review]

**Review by georgebisbas — COMMENTED** [review]
I guess no tests are needed for this PR for this new functionality?

**Review by mloubout — APPROVED** [review]

File: devito/types/parallel.py (hunk 4 above)
Comment (FabioLuporini) [reply devito/types/parallel.py:263]: yes, agree with Carlos here. We can't just return a (mutable) global value from `default_value` :-)

**Pull request merged**